### PR TITLE
Implement 'requestPermissions'

### PR DESCRIPTION
**Pull request by dpgraham**
(see https://developer.android.com/reference/androidx/test/runner/permission/PermissionRequester)

### CODE REVIEW

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ScreenshotHandler.java [hunk 1]
```diff
@@ -16,15 +16,25 @@
...
+
+import androidx.test.runner.screenshot.ScreenCapture;
+import androidx.test.runner.screenshot.Screenshot;
 import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
-import io.appium.espressoserver.lib.helpers.ScreenshotsHelper;
 import io.appium.espressoserver.lib.model.AppiumParams;
-import io.appium.espressoserver.lib.viewaction.ViewGetter;
 
-public class Screenshot implements RequestHandler<AppiumParams, String> {
+public class ScreenshotHandler implements RequestHandler<AppiumParams, String> {
```
Comment (jlipps) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/ScreenshotHandler.java]: what does this change have to do with permissions?

File: lib/commands/general.js [hunk 2]
```diff
@@ -42,6 +42,18 @@ commands.mobileIsToastVisible = async function (opts = {}) {
   });
 };
 
+commands.mobileRequestPermissions = async function (opts = {}) {
```
Comment (jlipps) [inline lib/commands/general.js]: why make this a mobile: method? i'd prefer to use the `permissions` cap we're already using for ios. or if we want permissions to be granted mid-run, use the appium settings API.

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import io.appium.espressoserver.lib.model.RequestPermissionsParams;
+
+public class RequestPermissionsHandler implements RequestHandler<RequestPermissionsParams, Void> {
+    @Override
+    public Void handle(RequestPermissionsParams params) throws AppiumException {
+        final PermissionRequester permissionRequester = new PermissionRequester();
+        for (final String permission: params.getPermissions()) {
+            try {
+                permissionRequester.addPermissions(permission);
+            } catch (Throwable e) {
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java]: It is not recommended to catch Throwable. Replace it with Exception

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                throw new AppiumException(String.format("Could not grant permission '%s'. Reason: %s", permission, e.getMessage()));
```
Comment (mykola-mokhnach) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java]: I'd add the original exception as `cause`

File: lib/commands/general.js [hunk 5]
```diff
@@ -42,6 +42,18 @@ commands.mobileIsToastVisible = async function (opts = {}) {
...
+  let {permissions} = opts;
+  if (!util.hasValue(permissions)) {
+
```
Comment (mykola-mokhnach) [inline lib/commands/general.js]: why it is empty?

File: lib/commands/general.js (hunk 2 above)
Comment (mykola-mokhnach) [reply lib/commands/general.js]: It probably makes sense to grant them in runtime as well, but the method should be more universal, so we can do both - grant and revoke a particular permission.

Comment (dpgraham) [reply lib/commands/general.js]: I couldn't find an API method that revokes permissions.

Comment (mykola-mokhnach) [reply lib/commands/general.js]: we can still do it via adb. androidx also contains helpers to execute shell commands

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.model.RequestPermissionsParams;
+
+public class RequestPermissionsHandler implements RequestHandler<RequestPermissionsParams, Void> {
+    @Override
+    public Void handle(RequestPermissionsParams params) throws AppiumException {
+        final PermissionRequester permissionRequester = new PermissionRequester();
+        for (final String permission: params.getPermissions()) {
+            try {
+                permissionRequester.addPermissions(permission);
```
Comment (KazuCocoa) [inline espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java]: What about defending  `params.getPermissions()` as `String[]`?
Then, we can get rid of `for` and can call ` permissionRequester.addPermissions(params.getPermissions());`, simply.

Comment (dpgraham) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java]: `addPermissions` accepts Strings... I don't think it accepts `String[]`

Comment (KazuCocoa) [reply espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/RequestPermissionsHandler.java]: ah, I have simply implemented the same code as:

```kotlin
    private fun grantPermissions(permissions: Array<String>) {
        val requester = PermissionRequester()
        requester.addPermissions(*permissions)
        requester.requestPermissions()
    }
```

The converted to Java code was:

```java
   private final void grantPermissions(String[] permissions) {
      PermissionRequester requester = new PermissionRequester();
      requester.addPermissions((String[])Arrays.copyOf(permissions, permissions.length));
      requester.requestPermissions();
   }
```

So I thought so.

**Comment by dpgraham** [comment]
I'm going to close this for now. I won't have time to prioritize this issue, and by the sounds of it we already have this functionality through ADB.

**Pull request closed without merging**